### PR TITLE
MAINT: Fix Windows build warnings in `linalg.interpolative`

### PR DESCRIPTION
**Pull request by Smit-create**
* Reference issue: https://github.com/rgommers/scipy/issues/120

cc @rgommers

### CODE REVIEW

**Comment by ilayn** [comment]
I'm a bit on the fence about whether we should provide feedback to Mark Tygert and get a new version or not. But changes are mostly conformal changes so I guess we can live with it.

**Comment by rgommers** [comment]
I had to refresh my memory on the history of it. @ilayn I believe this is completely unmaintained code, not even under version control in http://tygert.com/software.html, and hasn't changed in many years. So I wouldn't bother.

**Comment by rgommers** [comment]
@Smit-create Linux tests are failing with incorrect results for the routines you changed.

**Comment by Smit-create** [comment]
This failure seems unrelated:
```
FAILED ../testenv/lib/python3.10/site-packages/scipy/stats/tests/test_distributions.py::TestTruncnorm::test_moments
```

**Comment by WarrenWeckesser** [comment]
> This failure seems unrelated:
> 
> ```
> FAILED ../testenv/lib/python3.10/site-packages/scipy/stats/tests/test_distributions.py::TestTruncnorm::test_moments
> ```

That failure is sporadic; see https://github.com/scipy/scipy/issues/15871.

**Comment by rgommers** [comment]
@Smit-create did this actually help, or are we now getting the same warnings but the other way around? The CI log shows:
```
[74/1564] Compiling Fortran object scipy/linalg/_interpolative.cp39-win_amd64.pyd.p/src_id_dist_src_dfft.f.obj
../scipy/linalg/src/id_dist/src/dfft.f:2902:72:

 2902 |       CALL DFFTI1 (N,WSAVE(N+1),WSAVE(2*N+1))
      |                                                                        1
Warning: Type mismatch in argument 'ifac' at (1); passed REAL(8) to INTEGER(4)
../scipy/linalg/src/id_dist/src/dfft.f:2835:72:
```

Before it said `... passed INTEGER(4) to REAL(8)` for some of these; it's hard to review if those are the same or not. Maybe you can post a diff of CI logs, or suggest how best to review?

The `ind(n)` change looks suspicious, because it's documented to be an indexing array, which is naturally an array of integers not reals.

**Comment by rgommers** [comment]
I have merged the whitespace changes in commit 717ad33bbc8. So this PR should be easier to review now.

**Comment by Smit-create** [comment]
```
2902 |       CALL DFFTI1 (N,WSAVE(N+1),WSAVE(2*N+1))
      |                                                                        1
 Warning: Type mismatch in argument 'ifac' at (1); passed REAL(8) to INTEGER(4)
```
I didn't make changes to this because CI was failing.


> The `ind(n)` change looks suspicious, because it's documented to be an indexing array, which is naturally an array of integers not reals.

Yeah you're right, I looked at some docs and it shows `ind` as indexing array, fixing some warnings of `INTEGER to REAL` then gives rise to some other warnings saying `REAL to INTEGER`. Maybe we can leave `ind` for now.

**Comment by rgommers** [comment]
> Maybe we can leave `ind` for now.

Yes, let's do that.

Can you also confirm that the other changes don't introduce new warnings? Maybe post the CI output on `main` and on this PR for the `interpolative` build?

File: scipy/linalg/src/id_dist/src/iddp_rsvd.f [hunk 1]
```diff
@@ -195,9 +195,9 @@ subroutine iddp_rsvd0(m,n,matvect,p1t,p2t,p3t,p4t,
 c       for further documentation).
 c
         implicit none
-        integer m,n,krank,list(n),ier
+        integer m,n,krank,ier
         real*8 p1t,p2t,p3t,p4t,p1,p2,p3,p4,u(m,krank),v(n,krank),
-     1         s(krank),proj(krank,n-krank),col(m*krank),
+     1         s(krank),proj(krank,n-krank),col(m*krank),list(n),
      2         work((krank+1)*(m+3*n)+26*krank**2)
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/iddp_rsvd.f:201]: Warning fixed:
```
./scipy/linalg/src/id_dist/src/iddp_rsvd.f:155:72:

  155 |      3                    w(iwork))
      |                                                                        1
Warning: Type mismatch in argument 'list' at (1); passed REAL(8) to INTEGER(4)
```

This disappears on this diff.

File: scipy/linalg/src/id_dist/src/iddp_asvd.f [hunk 2]
```diff
@@ -160,8 +160,8 @@ subroutine iddp_asvd0(m,n,a,krank,list,proj,u,v,s,ier,
 c       for further documentation).
 c
         implicit none
-        integer m,n,krank,list(n),ier
-        real*8 a(m,n),u(m,krank),v(n,krank),
+        integer m,n,krank,ier
+        real*8 a(m,n),u(m,krank),v(n,krank),list(n),
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/iddp_asvd.f:164]: Warning fixed:
```
../scipy/linalg/src/id_dist/src/iddr_asvd.f:74:72:

   74 |      1                  w(ilist),w(iproj),w(icol),w(iwork))
      |                                                                        1
Warning: Type mismatch in argument 'list' at (1); passed REAL(8) to INTEGER(4)
[79/1564] Compiling Fortran object scipy/linalg/_interpolative.cp39-win_amd64.pyd.p/src_id_dist_src_iddp_asvd.f.obj
../scipy/linalg/src/id_dist/src/iddp_asvd.f:121:72:

  121 |      1                    w(iui),w(ivi),w(isi),ier,w(icol),w(iwork))
      |                                                                        1
```
This disappears on this diff.

File: scipy/linalg/src/id_dist/src/iddr_rsvd.f [hunk 3]
```diff
@@ -124,9 +124,9 @@ subroutine iddr_rsvd0(m,n,matvect,p1t,p2t,p3t,p4t,
 c       for further documentation).
 c
         implicit none
-        integer m,n,krank,list(n),ier,k
+        integer m,n,krank,ier,k
         real*8 p1t,p2t,p3t,p4t,p1,p2,p3,p4,u(m,krank),v(n,krank),
-     1         s(krank),proj(krank*(n-krank)),col(m*krank),
+     1         s(krank),proj(krank*(n-krank)),col(m*krank),list(n),
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/iddr_rsvd.f:129]: Warning fixed:
```
[94/1564] Compiling Fortran object scipy/linalg/_interpolative.cp39-win_amd64.pyd.p/src_id_dist_src_iddr_rsvd.f.obj
../scipy/linalg/src/id_dist/src/iddr_rsvd.f:109:72:

  109 |      2                  w(ilist),w(iproj),w(icol),w(iwork))
      |                                                                        1
Warning: Type mismatch in argument 'list' at (1); passed REAL(8) to INTEGER(4)
```

File: scipy/linalg/src/id_dist/src/idzp_aid.f [hunk 4]
```diff
@@ -170,8 +170,8 @@ subroutine idzp_aid1(eps,n2,n,kranki,proj,krank,list,rnorms)
 c
         implicit none
         integer n,n2,kranki,krank,list(n),j,k
-        real*8 eps,rnorms(n)
-        complex*16 proj(n2*n)
+        real*8 eps
+        complex*16 proj(n2*n),rnorms(n)
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/idzp_aid.f:174]: Fixed:
```
../scipy/linalg/src/id_dist/src/idzp_aid.f:88:72:

   88 |      1                                   proj(m*n+1))
      |                                                                        1
Warning: Type mismatch in argument 'rnorms' at (1); passed COMPLEX(8) to REAL(8)
../scipy/linalg/src/id_dist/src/idzp_aid.f:91:72:

   91 |      1                                   krank,list,proj(n2*n+1))
      |                                                                        1
Warning: Type mismatch in argument 'rnorms' at (1); passed COMPLEX(8) to REAL(8)
```

File: scipy/linalg/src/id_dist/src/idzp_asvd.f [hunk 5]
```diff
@@ -159,9 +159,9 @@ subroutine idzp_asvd0(m,n,a,krank,list,proj,u,v,s,ier,
 c       for further documentation).
 c
         implicit none
-        integer m,n,krank,list(n),ier
+        integer m,n,krank,ier
         real*8 s(krank)
-        complex*16 a(m,n),u(m,krank),v(n,krank),
+        complex*16 a(m,n),u(m,krank),v(n,krank),list(n),
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/idzp_asvd.f:164]: Fixed:
```
../scipy/linalg/src/id_dist/src/idzp_asvd.f:122:72:

  122 |      1                    w(iui),w(ivi),w(isi),ier,w(icol),w(iwork))
      |                                                                        1
Warning: Type mismatch in argument 'list' at (1); passed COMPLEX(8) to INTEGER(4)
```

File: scipy/linalg/src/id_dist/src/iddr_asvd.f [hunk 6]
```diff
@@ -88,8 +88,8 @@ subroutine iddr_asvd0(m,n,a,krank,winit,u,v,s,ier,
 c       for further documentation).
 c
         implicit none
-        integer m,n,krank,list(n),ier
-        real*8 a(m,n),u(m,krank),v(n,krank),s(krank),
+        integer m,n,krank,ier
+        real*8 a(m,n),u(m,krank),v(n,krank),s(krank),list(n),
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/iddr_asvd.f:92]: Fixed:
```
../scipy/linalg/src/id_dist/src/iddr_asvd.f:74:72:

   74 |      1                  w(ilist),w(iproj),w(icol),w(iwork))
      |                                                                        1
Warning: Type mismatch in argument 'list' at (1); passed REAL(8) to INTEGER(4)
```

File: scipy/linalg/src/id_dist/src/idzr_rsvd.f [hunk 7]
```diff
@@ -125,10 +125,10 @@ subroutine idzr_rsvd0(m,n,matveca,p1t,p2t,p3t,p4t,
...
 c
         implicit none
-        integer m,n,krank,list(n),ier,k
+        integer m,n,krank,ier,k
         real*8 s(krank)
         complex*16 p1t,p2t,p3t,p4t,p1,p2,p3,p4,u(m,krank),v(n,krank),
-     1             proj(krank*(n-krank)),col(m*krank),
+     1             proj(krank*(n-krank)),col(m*krank),list(n),
      2             work((krank+1)*(m+3*n+10)+9*krank**2)
         external matveca,matvec
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/idzr_rsvd.f:133]: Fixed:
```
../scipy/linalg/src/id_dist/src/idzr_rsvd.f:110:72:

  110 |      2                  w(ilist),w(iproj),w(icol),w(iwork))
      |                                                                        1
Warning: Type mismatch in argument 'list' at (1); passed COMPLEX(8) to INTEGER(4)
```

File: scipy/linalg/src/id_dist/src/idzr_asvd.f [hunk 8]
```diff
@@ -91,9 +91,9 @@ subroutine idzr_asvd0(m,n,a,krank,winit,u,v,s,ier,
 c       for further documentation).
 c
         implicit none
-        integer m,n,krank,list(n),ier
+        integer m,n,krank,ier
         real*8 s(krank)
-        complex*16 a(m,n),u(m,krank),v(n,krank),
+        complex*16 a(m,n),u(m,krank),v(n,krank),list(n),
```
Comment (Smit-create) [inline scipy/linalg/src/id_dist/src/idzr_asvd.f:96]: Fixed:
```
../scipy/linalg/src/id_dist/src/idzr_asvd.f:77:72:

   77 |      1                  w(ilist),w(iproj),w(icol),w(iwork))
      |                                                                        1
Warning: Type mismatch in argument 'list' at (1); passed COMPLEX(8) to INTEGER(4)
```

File: scipy/linalg/src/id_dist/src/idzp_asvd.f (hunk 5 above)
Comment (rgommers) [reply scipy/linalg/src/id_dist/src/idzp_asvd.f:164]: This one is not correct it looks like, the warning just changes to `Warning: Type mismatch in argument 's' at (1); passed COMPLEX(8) to REAL(8)`.

**Review by rgommers — COMMENTED** [review]
All changes except one now seem correct. Can you either fix or remove the one change that's incorrect @Smit-create?

File: scipy/linalg/src/id_dist/src/idzp_asvd.f (hunk 5 above)
Comment (Smit-create) [reply scipy/linalg/src/id_dist/src/idzp_asvd.f:164]: That is a different warning talking about line 163:

https://github.com/scipy/scipy/blob/6432ef2d9cc1f628ef64111bb63008f1cd794c9f/scipy/linalg/src/id_dist/src/idzp_asvd.f#L161-L165

What I fixed is the warning of `'list'` which was `COMPLEX` to `INTEGER`.

Comment (rgommers) [reply scipy/linalg/src/id_dist/src/idzp_asvd.f:164]: I looked at it again, this is definitely not correct. The first usage of `list` in this routine is `call idz_copycols(m,n,a,krank,list,col)`; `list` in that routine is also integer. 

The problem with this code is deeper, `w(*)` is a work array declared as `complex*16` but used for storing data of different types. Example on line 83 of this file:
```
 call idzp_aid(eps,m,n,a,winit,krank,w(ilist),w(iproj))
```
The last two parameters are both `w`; the first one should be integer and the second one complex.

In general this come is extremely poorly written. I think I'll revert this change, merge the rest of the fixes, and then we can silence the remaining warnings.

**Review by rgommers — COMMENTED** [review]
On further review, this indeed makes compiler warnings go away, _however_ I don't understand why. The `list` parameters clearly _are_ `integer` and not `real` or `complex`. 

An example for `iddp_rsvd.f`:

(1) the call on line 152-155 is:
```fortran
          call iddp_rsvd0(m,n,matvect,p1t,p2t,p3t,p4t,
     1                    matvec,p1,p2,p3,p4,krank,w(iui),w(ivi),
     2                    w(isi),ier,w(ilist),w(iproj),w(icol),
     3                    w(iwork))
```
with `w(8)` declared as `real*8`.

(2) the `list` parameter is an array of indices, and its first usage on line 207 is:
```fortran
        call idd_getcols(m,n,matvec,p1,p2,p3,p4,krank,list,col,work)
```

(3) looking at that `idd_getcols` implementation, we see that `list` is declared as `integer` there.

This code is just intentionally misusing `w`. I think the explanation of why that works at all is given in https://stackoverflow.com/questions/55384730/intentional-type-mismatch-in-fortran.

The things I don't understand are:
1. why this gives compiler warnings before but not after the change (since there's still a type mismatch, it's just moved to another call), and
2. why this gives warnings with gcc on Windows but not on Linux.

I'm inclined to not merge this, because it feels like this code has bigger issues and this isn't quite correct. Let me ping some folks with more Fortran knowledge and see what they think. @ev-br, @WarrenWeckesser, @czgdp1807 any thoughts?

**Comment by czgdp1807** [comment]
I think the behaviour you have described above is incorrect and most probably is a compiler bug. With `f90` files and `gfortran 11.0.1`, I receive warnings whenever I pass real array to a subroutine accepting integer arrays and vice-versa. See below,

```fortran
program debug
    real :: list(4)
    call real2integer(list)

contains

    subroutine real2integer(list)
        integer :: list(4)
    end subroutine

end program
```

**Warnings**

```bash
/Users/czgdp1807/scipy_project/debug/linalg_fortran.f90:3:27:

    3 |     call real2integer(list)
      |                           1
Error: Type mismatch in argument ‘list’ at (1); passed REAL(4) to INTEGER(4)
```

**Comment by ilayn** [comment]
I think they are playing the "clever" union memory hack as you would do in a C union 

```c
union foo {
    int x;
    char y;
} test;
```

The memory is shared but depending on which one you access the content is interpreted accordingly. So you can write `test.char = "W"` but then read it as an integer via `.x` property. So it looks indeed like a scratch register. Somehow compiler found out about the intention and not liking it. I also don't know why it is complaining in this mysterious fashion on Windows. Maybe datatypes have different sizes on Windows that compiler can figure it out as a problem.

**Comment by rgommers** [comment]
> Somehow compiler found out about the intention and not liking it. I also don't know why it is complaining in this mysterious fashion on Windows.

I think it should be complaining everywhere. `complex*16 w(*)` may be a convenient way to allocate the right size of temporary memory, but it should then have casts in the correct places, right? At least in C/C++ there must be casts, not sure what Fortran is supposed to do here. Ideally the whole thing should be rewritten because it's bad practice it looks like, but short of that what's the correct way to interpret memory with the deliberately incorrect type such that modern compilers are happy?

**Comment by ilayn** [comment]
I'm closing this in favor of #20558 but @Smit-create a big thank you from me for even attempting to tame this code. It was seriously a challenging thing to figure things out.

**Comment by ilayn** [comment]
For posterity, I asked the same weird usage of real arrays mixed with integers on https://fortran-lang.discourse.group/t/passing-real8-array-variable-as-2d-integer4-array/ and user *PierU* kindly offered an explanation 

>Note that real*8 and integer*4 are non-standard declarations. The standard says that a real or integer variable shall occupy one numeric storage unit (without mentioning any size), while a double precision variable shall occupy exactly 2 numeric storage unit.

> In practice, with virtually all existing compilers, the numeric storage unit is 4 bytes.So, a real*8 element can be storage associated to 2 integer elements. But there is absolutely no casting of the content.

> Passing arguments with type mismatchs has always been illegal in Fortran, but it was a relatively common practice because 1) it was convenient (for instance to store indifferently integer or real variable in a work array, 2) it was guaranteed to work, and 3) there was no compile-time checks for that.

**Pull request closed without merging**